cmd/clair/main.go

```diff
@@ -98,88 +99,105 @@ func main() {
+		fail = true
+	}
+}
 
-	// HTTP API server goroutine.
-	srvs.Go(func() error {
-		zlog.Info(srvctx).Msg("launching http transport")
-		srvs, err := initialize.Services(srvctx, &conf)
+func serveAPI(ctx context.Context, cfg *config.Config) func() error {
+	return func() error {
+		zlog.Info(ctx).Msg("launching http transport")
+		srvs, err := initialize.Services(ctx, cfg)
 		if err != nil {
 			return fmt.Errorf("service initialization failed: %w", err)
 		}
 		srv := http.Server{
-			BaseContext: func(_ net.Listener) context.Context { return srvctx },
+			BaseContext: func(_ net.Listener) context.Context {
+				return context.WithoutCancel(ctx)
+			},
 		}
-		srv.Handler, err = httptransport.New(srvctx, &conf, srvs.Indexer, srvs.Matcher, srvs.Notifier)
+		srv.Handler, err = httptransport.New(ctx, cfg, srvs.Indexer, srvs.Matcher, srvs.Notifier)
 		if err != nil {
 			return fmt.Errorf("http transport configuration failed: %w", err)
 		}
-		l, err := net.Listen("tcp", conf.HTTPListenAddr)
+		l, err := net.Listen("tcp", cfg.HTTPListenAddr)
 		if err != nil {
 			return fmt.Errorf("http transport configuration failed: %w", err)
 		}
-		if conf.TLS != nil {
-			cfg, err := conf.TLS.Config()
+		if cfg.TLS != nil {
+			cfg, err := cfg.TLS.Config()
 			if err != nil {
 				return fmt.Errorf("tls configuration failed: %w", err)
 			}
 			cfg.NextProtos = []string{"h2"}
 			srv.TLSConfig = cfg
 			l = tls.NewListener(l, cfg)
 		}
-		down.Add(&srv)
 		health.Ready()
-		if err := srv.Serve(l); err != http.ErrServerClosed {
-			return fmt.Errorf("http transport failed to launch: %w", err)
-		}
-		return nil
-	})
 
-	// Signal handler goroutine.
-	go func() {
-		ctx, stop := signal.NotifyContext(ctx, os.Interrupt)
-		defer func() {
-			// Note that we're using a background context here, so that we get a
-			// full timeout if the signal handler has fired.
-			tctx, done := context.WithTimeout(context.Background(), 10*time.Second)
-			err := down.Shutdown(tctx)
-			if err != nil {
-				zlog.Error(ctx).Err(err).Msg("error shutting down server")
+		var eg errgroup.Group
+		eg.Go(func() error {
+			if err := srv.Serve(l); !errors.Is(err, http.ErrServerClosed) {
+				return fmt.Errorf("http transport failed to launch: %w", err)
 			}
-			done()
-			stop()
-			zlog.Info(ctx).Msg("unregistered signal handler")
-		}()
-		zlog.Info(ctx).Msg("registered signal handler")
-		select {
-		case <-ctx.Done():
-			zlog.Info(ctx).Stringer("signal", os.Interrupt).Msg("gracefully shutting down")
-		case <-srvctx.Done():
+			return nil
+		})
+		eg.Go(func() error {
+			<-ctx.Done()
+			ctx, done := context.WithTimeoutCause(context.Background(), 10*time.Second, context.Cause(ctx))
+			defer done()
+			return srv.Shutdown(ctx)
+		})
+		return eg.Wait()
+	}
+}
+
+func serveIntrospection(ctx context.Context, cfg *config.Config) func() error {
+	return func() error {
+		zlog.Info(ctx).Msg("launching introspection server")
+		srv, err := introspection.New(ctx, cfg, nil)
```

Comment:
What is the reason for having the healthcheck arg when we only pass nil, just placeholding for future?

Reply:
Yeah, it's a placeholder for some real healthcheck scheme (#1974, perhaps?).